### PR TITLE
[Sdk 366] SDK Internal Limits and unit tests added

### PR DESCRIPTION
**Pull request by ZahidZafar**

### CODE REVIEW

File: countlyCommon/countlyCommon/Entities/EntityBase/CountlyUserDetailsBase.cs [hunk 1]
```diff
@@ -271,13 +266,15 @@ public CustomInfo Custom
 
         [JsonIgnore]
         internal bool isNotified { get; set; }
+        internal bool isNotificationEnabled { get; set; }
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/Entities/EntityBase/CountlyUserDetailsBase.cs:269]: Add the "ignore" tag to this.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
These test files should also be added to flavours other than 3.5 and 4.0.

Move the trimming calls to a utility function.

**Comment by ZahidZafar** [comment]
> These test files should also be added to flavours other than 3.5 and 4.0.
> 
> Move the trimming calls to a utility function.
trim calls need configuration object, configuration object is part of 'Countly' instance. So to move these call in utility functions I will have to pass configuration object as well.
we have discussed it before as well.

**Comment by ArtursKadikis** [comment]
The trimming calls are still not moved to a utility class.

**Comment by ZahidZafar** [comment]
I left a comment, please read and left resonse.

File: countlyCommon/countlyCommon/CountlyBase.cs [hunk 2]
```diff
@@ -631,6 +688,36 @@ internal async Task<bool> RecordExceptionInternal(string error, string stackTrac
             }
         }
 
+
+        private string ManipulateStackTrace(string stackTrace)
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/CountlyBase.cs:692]: This has not been moved to the utility class.

File: countlyCommon/countlyCommon/CountlyBase.cs [hunk 3]
```diff
@@ -401,7 +422,19 @@ public static Task<bool> RecordEvent(string Key, int Count, double? Sum)
...
+
+            CountlyConfig config = Countly.Instance.Configuration;
+            if (Key.Length > config.MaxKeyLength) {
+                UtilityHelper.CountlyLogging("[CountlyBase] RecordEvent : Max allowed key length is " + Countly.Instance.Configuration.MaxKeyLength);
+                Key = Key.Substring(0, Countly.Instance.Configuration.MaxKeyLength);
+            }
+
+            Segmentation segments = UtilityHelper.RemoveExtraSegments(Segmentation, config.MaxSegmentationValues);
+            segments = UtilityHelper.FixSegmentKeysAndValues(segments, config.MaxKeyLength, config.MaxValueSize);
+            ;
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/CountlyBase.cs:435]: This is a useless semicolon

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
"UserDetailTests .cs" has not only a whitespace in it it's name, but the name is also missing a letter. It should be "UserDetailsTests"

No code formatting has been performed on these changes.

Throughout these changes you have added a lot of double empty lines. Replace them with a single empty line.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Key trimming should be moved to a common place

**Comment by ZahidZafar** [comment]
> Key trimming should be moved to a common place

We can move 'Key trimming' logic into 'RecordInternEvent' method or into a new method.
One more thing if user set the 'MaxKeyLenght' less than predefine internal event keys e.g '[CLY]_star_rating', then it will get trim and internal events  feature will not work accordingly. (edited)

**Pull request merged**